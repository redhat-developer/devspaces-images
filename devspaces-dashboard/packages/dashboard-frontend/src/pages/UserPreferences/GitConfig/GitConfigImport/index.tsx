/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

export const fieldId = 'git-configuration';
export const fileNamePlaceholder = 'Upload the .gitconfig file';
export const textAreaPlaceholder = 'Or paste the Git Configuration here';

import { FileUpload, FileUploadProps, ValidatedOptions } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  content: string | undefined;
  validated: ValidatedOptions;
  onChange: (content: string, isUpload: boolean) => void;
};

export type State = {
  content: string | undefined;
  filename: string | undefined;
  file: File | undefined;
  isLoading: boolean;
};

/**
 * A component that allows the user to upload the .gitconfig file.
 */
export class GitConfigImport extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      content: props.content,
      filename: undefined,
      file: undefined,
      isLoading: false,
    };
  }

  private handleFileInputChange(_event: React.ChangeEvent | React.DragEvent, file: File): void {
    this.setState({
      file,
      filename: file.name,
    });
  }

  private handleClearClick(): void {
    this.setState({
      content: undefined,
      file: undefined,
      filename: undefined,
    });
    this.props.onChange('', false);
  }

  private handleDataChange(content: string): void {
    this.setState({ content });
    this.props.onChange(content, true);
  }

  private handleTextChange(content: string): void {
    this.setState({
      content,
      file: undefined,
      filename: undefined,
    });
    this.props.onChange(content, false);
  }

  public render(): React.ReactElement {
    const { validated } = this.props;
    const { content, filename, isLoading } = this.state;

    const fileUploadValidated: FileUploadProps['validated'] =
      validated === ValidatedOptions.warning ? ValidatedOptions.success : validated;

    return (
      <FileUpload
        id={fieldId}
        data-testid={fieldId}
        value={content}
        filename={filename}
        filenamePlaceholder={fileNamePlaceholder}
        onFileInputChange={(event, file) => this.handleFileInputChange(event, file)}
        onClearClick={() => this.handleClearClick()}
        onDataChange={data => this.handleDataChange(data)}
        onTextChange={text => this.handleTextChange(text)}
        browseButtonText="Upload"
        isLoading={isLoading}
        isRequired={true}
        isReadOnly={false}
        validated={fileUploadValidated}
        type="text"
        onReadStarted={() => this.setState({ isLoading: true })}
        onReadFinished={() => this.setState({ isLoading: false })}
        onReadFailed={() => this.setState({ isLoading: false })}
        textAreaPlaceholder={textAreaPlaceholder}
        allowEditingUploadedText={true}
        hideDefaultPreview={false}
      />
    );
  }
}
