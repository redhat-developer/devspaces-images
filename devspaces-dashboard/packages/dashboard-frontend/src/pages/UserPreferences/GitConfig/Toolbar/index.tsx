/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Button,
  ButtonVariant,
  Toolbar,
  ToolbarContent,
  ToolbarItem,
} from '@patternfly/react-core';
import { PlusCircleIcon } from '@patternfly/react-icons';
import React from 'react';

export type Props = {
  onAdd: () => void;
};

export class GitConfigToolbar extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    return (
      <Toolbar style={{ paddingBottom: 0 }}>
        <ToolbarContent>
          <ToolbarItem alignment={{ default: 'alignRight' }}>
            <Button
              variant={ButtonVariant.link}
              icon={<PlusCircleIcon />}
              iconPosition="left"
              onClick={() => this.props.onAdd()}
            >
              Import Git Configuration
            </Button>
          </ToolbarItem>
        </ToolbarContent>
      </Toolbar>
    );
  }
}
