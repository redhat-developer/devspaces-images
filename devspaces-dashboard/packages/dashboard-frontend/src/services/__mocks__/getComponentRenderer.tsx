/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

export * from '@testing-library/react';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
