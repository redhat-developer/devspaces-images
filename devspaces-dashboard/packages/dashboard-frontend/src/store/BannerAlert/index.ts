/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Action, Reducer } from 'redux';
import { AppThunk } from '..';
import { createObject } from '../helpers';

export interface State {
  messages: string[];
}

export interface AddBannerAction {
  type: 'ADD_BANNER';
  message: string;
}

export interface RemoveBannerAction {
  type: 'REMOVE_BANNER';
  message: string;
}

type KnownAction = AddBannerAction | RemoveBannerAction;

export type ActionCreators = {
  addBanner: (message: string) => AppThunk<AddBannerAction>;
  removeBanner: (message: string) => AppThunk<RemoveBannerAction>;
};

export const actionCreators: ActionCreators = {
  addBanner:
    (message: string): AppThunk<AddBannerAction> =>
    dispatch => {
      dispatch({
        type: 'ADD_BANNER',
        message,
      });
    },

  removeBanner:
    (message: string): AppThunk<RemoveBannerAction> =>
    dispatch => {
      dispatch({
        type: 'REMOVE_BANNER',
        message,
      });
    },
};

const unloadedState: State = {
  messages: [],
};

export const reducer: Reducer<State> = (
  state: State | undefined,
  incomingAction: Action,
): State => {
  if (state === undefined) {
    return unloadedState;
  }

  const action = incomingAction as KnownAction;

  switch (action.type) {
    case 'ADD_BANNER':
      return createObject(state, {
        messages: state.messages.includes(action.message)
          ? state.messages
          : state.messages.concat([action.message]),
      });
    case 'REMOVE_BANNER':
      return createObject(state, {
        messages: state.messages.includes(action.message)
          ? state.messages.filter(message => message !== action.message)
          : state.messages,
      });
    default:
      return state;
  }
};
