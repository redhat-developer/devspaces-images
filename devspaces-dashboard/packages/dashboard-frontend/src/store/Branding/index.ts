/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Action, Reducer } from 'redux';
import { fetchBranding } from '../../services/assets/branding';
import { AppThunk } from '..';
import { merge } from 'lodash';
import axios from 'axios';
import common from '@eclipse-che/common';
import { BRANDING_DEFAULT, BrandingData } from '../../services/bootstrap/branding.constant';
import { container } from '../../inversify.config';
import { CheWorkspaceClient } from '../../services/workspace-client/cheworkspace/cheWorkspaceClient';
import { createObject } from '../helpers';
import { isSafari } from '../../services/helpers/detectBrowser';

const ASSET_PREFIX = './assets/branding/';

export interface State {
  isLoading: boolean;
  data: BrandingData;
  error?: string;
}

export interface RequestBrandingAction {
  type: 'REQUEST_BRANDING';
}

export interface ReceivedBrandingAction {
  type: 'RECEIVED_BRANDING';
  data: BrandingData;
}

export interface ReceivedBrandingErrorAction {
  type: 'RECEIVED_BRANDING_ERROR';
  error: string;
}

type KnownAction = RequestBrandingAction | ReceivedBrandingAction | ReceivedBrandingErrorAction;

export type ActionCreators = {
  requestBranding: () => AppThunk<KnownAction, Promise<void>>;
};

const cheWorkspaceClient = container.get(CheWorkspaceClient);

export const actionCreators: ActionCreators = {
  requestBranding:
    (): AppThunk<KnownAction, Promise<void>> =>
    async (dispatch): Promise<void> => {
      const url = `${ASSET_PREFIX}product.json`;

      dispatch({
        type: 'REQUEST_BRANDING',
      });

      let branding: BrandingData;
      try {
        const receivedBranding = await fetchBranding(url);
        branding = getBrandingData(receivedBranding);
        dispatch({
          type: 'RECEIVED_BRANDING',
          data: branding,
        });
      } catch (e) {
        const errorMessage =
          `Failed to fetch branding data by URL: "${url}", reason: ` +
          common.helpers.errors.getMessage(e);
        dispatch({
          type: 'RECEIVED_BRANDING_ERROR',
          error: errorMessage,
        });
        throw errorMessage;
      }

      // Use the products version if specified in product.json, otherwise use the default version given by che server
      if (!branding.productVersion) {
        try {
          const apiInfo = await getApiInfo();
          branding.productVersion = apiInfo.implementationVersion;

          dispatch({
            type: 'RECEIVED_BRANDING',
            data: branding,
          });
        } catch (e) {
          const errorMessage =
            'OPTIONS request to "/api/" failed, reason: ' + common.helpers.errors.getMessage(e);
          dispatch({
            type: 'RECEIVED_BRANDING_ERROR',
            error: errorMessage,
          });
          throw errorMessage;
        }
      }
    },
};

const unloadedState: State = {
  isLoading: false,
  data: getBrandingData(),
};

export const reducer: Reducer<State> = (
  state: State | undefined,
  incomingAction: Action,
): State => {
  if (state === undefined) {
    return unloadedState;
  }

  const action = incomingAction as KnownAction;
  switch (action.type) {
    case 'REQUEST_BRANDING':
      return createObject(state, {
        isLoading: true,
        error: undefined,
      });
    case 'RECEIVED_BRANDING':
      return createObject(state, {
        isLoading: false,
        data: action.data,
      });
    case 'RECEIVED_BRANDING_ERROR':
      return createObject(state, {
        isLoading: false,
        error: action.error,
      });
    default:
      return state;
  }
};

async function getApiInfo(): Promise<{ implementationVersion: string }> {
  if (isSafari) {
    return (await axios.options('/api/')).data;
  } else {
    return cheWorkspaceClient.restApiClient.getApiInfo();
  }
}

function getBrandingData(receivedBranding?: { [key: string]: any }): BrandingData {
  let branding: BrandingData = Object.assign({}, BRANDING_DEFAULT);

  if (receivedBranding && Object.keys(receivedBranding).length > 0) {
    branding = merge(branding, receivedBranding);
  }
  // resolve asset paths
  const assetTitles: Array<keyof BrandingData> = ['logoFile', 'logoTextFile'];
  assetTitles.forEach((asset: string) => {
    const path = branding[asset] as string;
    if (path.startsWith(ASSET_PREFIX)) {
      return;
    }
    branding[asset] = ASSET_PREFIX + branding[asset];
  });

  return branding;
}
