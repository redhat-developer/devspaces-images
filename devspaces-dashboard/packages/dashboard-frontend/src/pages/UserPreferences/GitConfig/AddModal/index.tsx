/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Button, ButtonVariant, Modal, ModalVariant } from '@patternfly/react-core';
import isEqual from 'lodash/isEqual';
import React from 'react';

import { GitConfigForm } from '@/pages/UserPreferences/GitConfig/AddModal/GitConfigForm';
import * as GitConfigStore from '@/store/GitConfig';

export type Props = {
  gitConfig: GitConfigStore.GitConfig | undefined;
  isOpen: boolean;
  onSave: (gitConfig: GitConfigStore.GitConfig) => Promise<void>;
  onCloseModal: () => void;
};
export type State = {
  gitConfig: GitConfigStore.GitConfig | undefined;
  isSaveEnabled: boolean;
};

export class GitConfigAddModal extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitConfig: undefined,
      // initially disabled until something changes and form is valid
      isSaveEnabled: false,
    };
  }

  private handleSaveGitConfig(): void {
    const { gitConfig } = this.state;

    if (gitConfig) {
      this.props.onSave(gitConfig);
    }
  }

  private handleCloseModal(): void {
    this.props.onCloseModal();
  }

  private handleChangeGitConfig(
    gitConfig: GitConfigStore.GitConfig | undefined,
    isValid: boolean,
  ): void {
    const isSaveEnabled =
      isValid && gitConfig !== undefined && !isEqual(this.props.gitConfig, gitConfig);
    this.setState({
      gitConfig,
      isSaveEnabled,
    });
  }

  private buildModalFooter(): React.ReactNode {
    return (
      <React.Fragment>
        <Button
          variant={ButtonVariant.primary}
          isDisabled={!this.state.isSaveEnabled}
          onClick={() => this.handleSaveGitConfig()}
        >
          Add
        </Button>
        <Button variant={ButtonVariant.link} onClick={() => this.handleCloseModal()}>
          Cancel
        </Button>
      </React.Fragment>
    );
  }

  public render(): React.ReactElement {
    const { isOpen } = this.props;

    const modalTitle = 'Import Git Configuration';
    const modalFooter = this.buildModalFooter();

    return (
      <Modal
        aria-label={modalTitle}
        title={modalTitle}
        variant={ModalVariant.small}
        isOpen={isOpen}
        onClose={() => this.handleCloseModal()}
        footer={modalFooter}
      >
        <GitConfigForm
          gitConfig={this.props.gitConfig}
          onChange={(gitConfig, isValid) => this.handleChangeGitConfig(gitConfig, isValid)}
        />
      </Modal>
    );
  }
}
