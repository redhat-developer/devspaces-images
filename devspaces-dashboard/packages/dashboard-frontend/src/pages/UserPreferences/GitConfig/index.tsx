/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { helpers } from '@eclipse-che/common';
import { AlertVariant, PageSection } from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import ProgressIndicator from '@/components/Progress';
import { lazyInject } from '@/inversify.config';
import { GitConfigAddModal } from '@/pages/UserPreferences/GitConfig/AddModal';
import { GitConfigEmptyState } from '@/pages/UserPreferences/GitConfig/EmptyState';
import { GitConfigForm } from '@/pages/UserPreferences/GitConfig/Form';
import { GitConfigToolbar } from '@/pages/UserPreferences/GitConfig/Toolbar';
import { AppAlerts } from '@/services/alerts/appAlerts';
import { AppState } from '@/store';
import * as GitConfigStore from '@/store/GitConfig';
import {
  selectGitConfig,
  selectGitConfigError,
  selectGitConfigIsLoading,
} from '@/store/GitConfig/selectors';

export type Props = MappedProps;

export type State = {
  isAddEditOpen: boolean;
};

class GitConfig extends React.PureComponent<Props, State> {
  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);

    this.state = {
      isAddEditOpen: false,
    };
  }

  public async componentDidMount(): Promise<void> {
    const { gitConfigIsLoading } = this.props;

    if (gitConfigIsLoading === true) {
      this.init();
      return;
    }

    await this.handleReload();
  }

  public componentDidUpdate(prevProps: Props): void {
    this.init(prevProps);
  }

  private init(prevProps?: Props): void {
    const { gitConfigError } = this.props;
    const prevGitConfigError = prevProps?.gitConfigError;
    if (gitConfigError && gitConfigError !== prevGitConfigError) {
      this.appAlerts.showAlert({
        key: 'gitconfig-error',
        title: helpers.errors.getMessage(gitConfigError),
        variant: AlertVariant.danger,
      });
    } else if (gitConfigError === undefined && gitConfigError !== prevGitConfigError) {
      this.appAlerts.removeAlert('gitconfig-error');
    }
  }

  private async handleSave(gitConfig: GitConfigStore.GitConfig): Promise<void> {
    try {
      await this.props.updateGitConfig(gitConfig);
      this.appAlerts.showAlert({
        key: 'gitconfig-success',
        title: 'Gitconfig saved successfully.',
        variant: AlertVariant.success,
      });
    } catch (error) {
      console.error('Failed to update gitconfig', error);
    }
  }

  private async handleReload(): Promise<void> {
    try {
      await this.props.requestGitConfig();
    } catch (error) {
      console.error('Failed to reload gitconfig', error);
    }
  }

  private handleCloseAddEditModal(): void {
    this.setState({
      isAddEditOpen: false,
    });
  }

  public render(): React.ReactElement {
    const { gitConfigIsLoading, gitConfig } = this.props;
    const { isAddEditOpen } = this.state;

    const isEmpty = gitConfig === undefined;

    return (
      <React.Fragment>
        <ProgressIndicator isLoading={gitConfigIsLoading} />
        <GitConfigAddModal
          gitConfig={gitConfig}
          isOpen={isAddEditOpen}
          onCloseModal={() => this.handleCloseAddEditModal()}
          onSave={async gitConfig => {
            await this.handleSave(gitConfig);
            this.handleCloseAddEditModal();
          }}
        />
        {isEmpty ? (
          <GitConfigEmptyState />
        ) : (
          <PageSection>
            <GitConfigToolbar
              onAdd={() => {
                this.setState({
                  isAddEditOpen: true,
                });
              }}
            />
            <GitConfigForm
              gitConfig={gitConfig}
              isLoading={gitConfigIsLoading}
              onSave={async gitConfig => await this.handleSave(gitConfig)}
              onReload={async () => await this.handleReload()}
            />
          </PageSection>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  gitConfig: selectGitConfig(state),
  gitConfigIsLoading: selectGitConfigIsLoading(state),
  gitConfigError: selectGitConfigError(state),
});

const connector = connect(mapStateToProps, GitConfigStore.actionCreators, null, {
  // forwardRef is mandatory for using `@react-mock/state` in unit tests
  forwardRef: true,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(GitConfig);
