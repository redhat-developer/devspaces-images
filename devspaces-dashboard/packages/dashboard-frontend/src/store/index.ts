/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';
import * as BannerAlertStore from './BannerAlert';
import * as BrandingStore from './Branding';
import * as ClusterInfo from './ClusterInfo';
import * as ClusterConfig from './ClusterConfig';
import * as DevfileRegistriesStore from './DevfileRegistries';
import * as FactoryResolverStore from './FactoryResolver';
import * as InfrastructureNamespacesStore from './InfrastructureNamespaces';
import * as PluginsStore from './Plugins/chePlugins';
import * as UserPreferences from './UserPreferences';
import * as WorkspacesStore from './Workspaces';
import * as CheWorkspacesStore from './Workspaces/cheWorkspaces';
import * as DevWorkspacesStore from './Workspaces/devWorkspaces';
import * as UserStore from './User';
import * as UserProfileStore from './UserProfile';
import * as DwPluginsStore from './Plugins/devWorkspacePlugins';
import * as WorkspacesSettingsStore from './Workspaces/Settings';
import * as CheDockerConfigStore from './DockerConfig/che';
import * as DwDockerConfigStore from './DockerConfig/dw';

// the top-level state object
export interface AppState {
  bannerAlert: BannerAlertStore.State;
  branding: BrandingStore.State;
  clusterInfo: ClusterInfo.State;
  clusterConfig: ClusterConfig.State;
  devfileRegistries: DevfileRegistriesStore.State;
  infrastructureNamespaces: InfrastructureNamespacesStore.State;
  user: UserStore.State;
  userProfile: UserProfileStore.State;
  workspaces: WorkspacesStore.State;
  cheWorkspaces: CheWorkspacesStore.State;
  devWorkspaces: DevWorkspacesStore.State;
  plugins: PluginsStore.State;
  factoryResolver: FactoryResolverStore.State;
  userPreferences: UserPreferences.State;
  dwPlugins: DwPluginsStore.State;
  workspacesSettings: WorkspacesSettingsStore.State;
  cheDockerConfig: CheDockerConfigStore.State;
  dwDockerConfig: DwDockerConfigStore.State;
}

export const reducers = {
  bannerAlert: BannerAlertStore.reducer,
  workspaces: WorkspacesStore.reducer,
  cheWorkspaces: CheWorkspacesStore.reducer,
  devWorkspaces: DevWorkspacesStore.reducer,
  devfileRegistries: DevfileRegistriesStore.reducer,
  branding: BrandingStore.reducer,
  clusterInfo: ClusterInfo.reducer,
  clusterConfig: ClusterConfig.reducer,
  user: UserStore.reducer,
  userProfile: UserProfileStore.reducer,
  infrastructureNamespaces: InfrastructureNamespacesStore.reducer,
  plugins: PluginsStore.reducer,
  factoryResolver: FactoryResolverStore.reducer,
  userPreferences: UserPreferences.reducer,
  dwPlugins: DwPluginsStore.reducer,
  workspacesSettings: WorkspacesSettingsStore.reducer,
  cheDockerConfig: CheDockerConfigStore.reducer,
  dwDockerConfig: DwDockerConfigStore.reducer,
};

export type AppThunk<ActionType extends Action, ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  ActionType
>;
