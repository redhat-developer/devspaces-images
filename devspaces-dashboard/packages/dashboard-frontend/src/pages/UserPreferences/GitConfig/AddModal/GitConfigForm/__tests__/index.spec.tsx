/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';

import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';
import * as GitConfigStore from '@/store/GitConfig';

import { GitConfigForm } from '..';

const mockOnChange = jest.fn();

jest.mock('@/pages/UserPreferences/GitConfig/GitConfigImport');

const { renderComponent, createSnapshot } = getComponentRenderer(getComponent);

describe('GitConfigForm', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot without predefined git configuration', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });
  test('snapshot with predefined git configuration', () => {
    const snapshot = createSnapshot({ user: { email: 'user-1@chetest.com', name: 'User One' } });
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should handle a valid value', () => {
    renderComponent();

    const gitConfigField = screen.getByTestId('submit-valid-git-config');
    userEvent.click(gitConfigField);

    expect(mockOnChange).toHaveBeenCalledWith(
      { user: { email: 'user-1@chetest.com', name: 'User One' } },
      true,
    );
  });

  it('should handle an invalid value', () => {
    renderComponent();

    const gitConfigField = screen.getByTestId('submit-invalid-git-config');
    userEvent.click(gitConfigField);

    expect(mockOnChange).toHaveBeenCalledWith({ user: { name: 'User One' } }, false);
  });
});

function getComponent(gitConfig?: GitConfigStore.GitConfig) {
  return <GitConfigForm onChange={mockOnChange} gitConfig={gitConfig} />;
}
