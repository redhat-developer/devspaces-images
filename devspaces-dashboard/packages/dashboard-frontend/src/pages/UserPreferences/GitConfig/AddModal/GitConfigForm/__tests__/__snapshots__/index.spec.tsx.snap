// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitConfigForm snapshot with predefined git configuration 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="gitconfig"
      >
        <span
          className="pf-c-form__label-text"
        >
          gitconfig
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <div>
        <input
          data-testid="submit-invalid-git-config"
          onClick={[Function]}
          type="button"
          value="[user]
    email="user-1@chetest.com"
    name="User One"
    "
        />
        <input
          data-testid="submit-valid-git-config"
          onClick={[Function]}
          type="button"
        />
      </div>
      
    </div>
  </div>
</form>
`;

exports[`GitConfigForm snapshot without predefined git configuration 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="gitconfig"
      >
        <span
          className="pf-c-form__label-text"
        >
          gitconfig
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <div>
        <input
          data-testid="submit-invalid-git-config"
          onClick={[Function]}
          type="button"
        />
        <input
          data-testid="submit-valid-git-config"
          onClick={[Function]}
          type="button"
        />
      </div>
      
    </div>
  </div>
</form>
`;
